WalletWasabi/EventSourcing/EventPubSub.cs

```diff
@@ -0,0 +1,71 @@
+using System.Linq;
+using System.Threading.Tasks;
+using WalletWasabi.EventSourcing.Interfaces;
+using WalletWasabi.EventSourcing.Records;
+
+namespace WalletWasabi.EventSourcing
+{
+	public class EventPubSub : IEventPubSub
+	{
+		#region Dependencies
+
+		protected IEventRepository EventRepository { get; init; }
+		protected IPubSub PubSub { get; init; }
+
+		#endregion Dependencies
+
+		public EventPubSub(IEventRepository eventRepository, IPubSub pubSub)
+		{
+			EventRepository = eventRepository;
+			PubSub = pubSub;
+		}
+
+		/// <inheritdoc/>
+		public async Task PublishAllAsync()
+		{
+			try
+			{
+				var aggregatesEvents = await EventRepository.ListUndeliveredEventsAsync().ConfigureAwait(false);
+				await aggregatesEvents.ForEachAggregateExceptionsAsync(
+					async (aggregateEvents) =>
+					{
+						if (0 < aggregateEvents.WrappedEvents.Count)
+						{
+							try
+							{
+								await aggregateEvents.WrappedEvents.ForEachAggregateExceptionsAsync(
+									PubSub.PublishDynamicAsync
+								).ConfigureAwait(false);
+							}
+							catch
+							{
+								// TODO: move events into tombstone queue for redelivery
+								// with exponential back-off with sprinkle of random delay
+								// and then mark those events as delivered in the event store
+								// to escape this loop of infinite redelivery attempts
+								throw;
```

Comment:
when might this become important? is it fair to say that exceptions in pubsub delivery currently imply bugs in their implementation, but this might change if we use some sort of IPC or networked bus at some point in the future? in that case would it just be another `catch` for ephemeral exceptions?

Reply:
well, even in case of a bug you still need to eventually fix your data. You deploy new code with a bug fix and the event is retried from the tombstone and data are fixed without any manual action. If you lose the event your data becomes inconsistent. It is a must for guaranteed eventual consistency.

Reply:
thanks, makes sense